File: sdk/php/src/Codegen/Introspection/AbstractVisitor.php
<?php

namespace Dagger\Codegen\Introspection;

use Dagger\Codegen\CodeWriter;
use GraphQL\Type\Definition\Type;
use GraphQL\Type\Schema;
use Nette\PhpGenerator\ClassType;
use Nette\PhpGenerator\EnumType;

abstract class AbstractVisitor extends CodeWriter
{
    public function __construct(protected readonly Schema $schema, string $targetDirectory)
    {
        parent::__construct($targetDirectory);
    }

    public function visit(Type $type): void
    {
        $this->write($this->generateType($type));
    }

    abstract public function generateType(Type $type): EnumType|ClassType;
}

File: sdk/java/dagger-codegen-maven-plugin/src/main/java/io/dagger/codegen/introspection/AbstractVisitor.java
package io.dagger.codegen.introspection;

import com.palantir.javapoet.TypeSpec;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Path;

abstract class AbstractVisitor extends CodeWriter {

  private Schema schema;

  public AbstractVisitor(Schema schema, Path targetDirectory, Charset encoding) {
    super(targetDirectory, encoding);
    this.schema = schema;
  }

  void visit(Type type) throws IOException {
    TypeSpec typeSpec = generateType(type);
    write(typeSpec);
  }

  public Schema getSchema() {
    return schema;
  }

  abstract TypeSpec generateType(Type type);
}
